crates/runtime/src/dataconnector/ftp.rs

```diff
@@ -0,0 +1,189 @@
+    UnableToBuildObjectStore {
+        source: object_store::Error,
+    },
+
+    ObjectStoreNotImplemented,
+
+    #[snafu(display("{source}"))]
+    UnableToBuildLogicalPlan {
+        source: DataFusionError,
+    },
+
+    #[snafu(display("Unsupported file format {format} in S3 Connector"))]
+    UnsupportedFileFormat {
+        format: String,
+    },
+
+    #[snafu(display("Unsupported compression type for CSV"))]
+    UnsupportedCompressionType {
+        source: DataFusionError,
+        compression_type: String,
+    },
+}
+
+pub struct FTP {
+    secret: Option<Secret>,
+    params: HashMap<String, String>,
+}
+
+impl DataConnectorFactory for FTP {
+    fn create(
+        secret: Option<Secret>,
+        params: Arc<Option<HashMap<String, String>>>,
+    ) -> Pin<Box<dyn Future<Output = super::NewDataConnectorResult> + Send>> {
+        Box::pin(async move {
+            let ftp = Self {
+                secret,
+                params: params.as_ref().clone().map_or_else(HashMap::new, |x| x),
+            };
+            Ok(Arc::new(ftp) as Arc<dyn DataConnector>)
+        })
+    }
+}
+
+impl FTP {
+    fn get_object_store_url(&self, dataset: &Dataset) -> AnyErrorResult<Url> {
+        let mut fragments = vec![];
+        let mut fragment_builder = form_urlencoded::Serializer::new(String::new());
+
+        if let Some(ftp_port) = self.params.get("ftp_port") {
+            fragment_builder.append_pair("port", ftp_port);
+        }
+        if let Some(ftp_user) = self.params.get("ftp_user") {
+            fragment_builder.append_pair("user", ftp_user);
+        }
+        if let Some(ftp_password) =
+            get_secret_or_param(Some(&self.params), &self.secret, "ftp_pass_key", "ftp_pass")
+        {
+            fragment_builder.append_pair("password", &ftp_password);
+        }
+        fragments.push(fragment_builder.finish());
+
+        let mut ftp_url =
+            Url::parse(&dataset.from).context(UnableToParseURLSnafu { url: &dataset.from })?;
+
+        if dataset.from.clone().ends_with('/') {
+            fragments.push("dfiscollectionbugworkaround=hack/".into());
+        }
+
+        ftp_url.set_fragment(Some(&fragments.join("&")));
+
+        Ok(ftp_url)
+    }
+
+    fn get_file_format_and_extension(&self) -> AnyErrorResult<(Arc<dyn FileFormat>, String)> {
+        let params = &self.params;
+        let extension = params.get("file_extension").cloned();
+
+        match params.get("file_format").map(String::as_str) {
+            Some("csv") => Ok((
+                get_csv_format(params)?,
+                extension.unwrap_or(".csv".to_string()),
+            )),
+            None | Some("parquet") => Ok((
+                Arc::new(ParquetFormat::default()),
+                extension.unwrap_or(".parquet".to_string()),
+            )),
+            Some(format) => Err(Error::UnsupportedFileFormat {
+                format: format.to_string(),
+            }
+            .into()),
+        }
+    }
+}
```

Comment:
this should be same as S3 one?

Reply:
Created new trait ListingTableConnector with auto implementation of DataConnector